File: openharmony-sig/ohos_react_native/tester/harmony/react_native_openharmony/src/main/cpp/RNOHCorePackage/TurboModules/I18nManagerTurboModule.cpp
/**
 * Copyright (c) 2024 Huawei Technologies Co., Ltd.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE-MIT file in the root directory of this source tree.
 */

#include "RNOHCorePackage/TurboModules/I18nManagerTurboModule.h"

using namespace rnoh;
using namespace facebook;

I18nManagerTurboModule::I18nManagerTurboModule(
    const ArkTSTurboModule::Context ctx,
    const std::string name)
    : ArkTSTurboModule(ctx, name) {
  methodMap_ = {
      ARK_METHOD_METADATA(getConstants, 0),
      ARK_SCHEDULE_METHOD_METADATA(allowRTL, 1),
      ARK_SCHEDULE_METHOD_METADATA(forceRTL, 1)};
}

File: openharmony-sig/ohos_react_native/tester/harmony/react_native_openharmony/src/main/ets/RNOHCorePackage/turboModules/I18nManagerTurboModule.ts
/**
 * Copyright (c) 2024 Huawei Technologies Co., Ltd.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE-MIT file in the root directory of this source tree.
 */

import I18n from '@ohos.i18n';
import { TurboModule } from "../../RNOH/TurboModule";

export class I18nManagerTurboModule extends TurboModule {
  public static readonly NAME = 'I18nManager';

  private RTLAllowed = true;
  private RTLForced = false;

  get isRTL() {
    return this.RTLForced || (this.RTLAllowed && I18n.isRTL(this.ctx.uiAbilityContext.config.language));
  }

  getConstants() {
    return {
      isRTL: this.isRTL,
      doLeftAndRightSwapInRTL: true,
      localeIdentifier: this.ctx.uiAbilityContext.config.language
    };
  }

  allowRTL(allow: boolean) {
    this.RTLAllowed = allow;
    this.ctx.rnInstance.updateRTL(this.isRTL);
  }

  forceRTL(force: boolean) {
    this.RTLForced = force;
    this.ctx.rnInstance.updateRTL(this.isRTL);
  }
}
